drivers: firmware: scmi: Introduce basic system power protocol

yongxu-wang15 opened the pull request:
Add SCMI System Power Domain Protocol support with:

- System power state set operations
- Protocol version/attributes query operations
- Standard SCMI power states (shutdown, cold/warm reset, suspend)
- Graceful and forceful power transitions
- Optional NXP vendor-specific power states
- Device tree binding for "arm,scmi-system"

This enables system-wide power management through standardized SCMI interfaces for reboot and power control operations.

yongxu-wang15 committed:
drivers: firmware: scmi: Introduce basic system power protocol

Add SCMI System Power Domain Protocol support with:

- System power state set operations
- Standard SCMI power states (shutdown, cold/warm reset, suspend)
- Graceful and forceful power transitions
- Optional NXP vendor-specific power states
- Device tree binding for "arm,scmi-system"

This enables system-wide power management through standardized
SCMI interfaces for reboot and power control operations.

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

dsseng commented on drivers/firmware/scmi/reboot.c:
```diff
@@ -0,0 +1,86 @@
+#include <zephyr/drivers/firmware/scmi/system.h>
+#include <string.h>
+
+#if defined(CONFIG_REBOOT)
```
`config ARM_SCMI_REBOOT` `depends on ARM_SCMI_SYSTEM && REBOOT`

Thus, this #if is always selected. I believe it should be safe to remove these conditions

dsseng reviewed (commented):
Thanks for sharing this code not only between SoCs but also generalizing it!

yongxu-wang15 committed:
drivers: firmware: scmi: add common protocol interface functions

Add standardized SCMI protocol common interface functions that can be
reused across different SCMI protocol implementations:

- scmi_protocol_get_version(): Query protocol version
- scmi_protocol_attributes_get(): Get protocol-specific attributes
- scmi_protocol_message_attributes_get(): Query message capabilities
- scmi_protocol_version_negotiate(): Negotiate protocol version support

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

yongxu-wang15 committed:
drivers: firmware: scmi: add initial reboot interface

Implement sys_arch_reboot() using SCMI system power protocol
for warm and cold reset operations on SCMI-enabled platforms.

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

yongxu-wang15 commented on drivers/firmware/scmi/reboot.c:
```diff
@@ -0,0 +1,86 @@
+		return 0;
+	}
+
+	LOG_INF("SCMI system protocol version: 0x%08x", version);
```
yes, its needed to add some version check and negotiate in protocol init phase, have added this flow in newly commit, please help to review it, thanks

dsseng commented on drivers/firmware/scmi/core.c:
```diff
@@ -215,6 +215,47 @@ int scmi_send_message(struct scmi_protocol *proto, struct scmi_message *msg,
+				LOG_ERR("Protocol 0x%X: Negotiation failed (%d)", proto->id, ret);
+				return ret;
+			}
+		} else {
```
The case when our version is higher than the platform's exposed version (firmware is too old) seems to falsely fall into the compatible category.

yongxu-wang15 commented on drivers/firmware/scmi/core.c:
```diff
@@ -215,6 +215,47 @@ int scmi_send_message(struct scmi_protocol *proto, struct scmi_message *msg,
+		ret = scmi_protocol_version_get(proto, &agent_version);
+		if (ret < 0) {
+			LOG_ERR("Protocol 0x%X: Failed to get agent version: %d",
+				proto->id, ret);
```
okay, updated into scmi_protocol_get_version, thanks

yongxu-wang15 commented on drivers/firmware/scmi/core.c:
```diff
@@ -215,6 +215,47 @@ int scmi_send_message(struct scmi_protocol *proto, struct scmi_message *msg,
+			LOG_INF("Protocol 0x%X: Platform version 0x%08x is compatible",
+				proto->id, agent_version);
+		}
+	}
```
sure, removed some duplicated LOG in there, only keep LOG_INF in last and LG_WRN if negotiate fail

yongxu-wang15 commented on drivers/firmware/scmi/common.c:
```diff
@@ -0,0 +1,148 @@
+};
+
+int scmi_protocol_version_get(struct scmi_protocol *proto, uint32_t *version)
+{
```
protocol.h file have added these api define description, plase check it in that, thanks

LaurentiuM1234 commented on drivers/firmware/scmi/common.c:
```diff
@@ -0,0 +1,148 @@
+
+	use_polling = k_is_pre_kernel();
+
+	ret = scmi_send_message(proto, &msg, &reply, use_polling);
```
can just use `k_is_pre_kernel()` directly here instead of `use_polling`?

yongxu-wang15 replied:
sure, its enough expectly scmi_protocol_message_attributes_get which used in reboot context, irq is diabled in system reboot level, so that apply poll mode for it in newly commit

LaurentiuM1234 commented on drivers/firmware/scmi/common.c:
```diff
@@ -0,0 +1,148 @@
+
+	use_polling = k_is_pre_kernel();
+
+	ret = scmi_send_message(proto, &msg, &reply, use_polling);
```
ditto

yongxu-wang15 committed:
drivers: firmware: scmi: add protocol version support

Add version field to struct scmi_protocol and update protocol macros
to accept version parameter.

Protocol versions added:
  - Clock Control: v3.0.0 (SCMI v3.2 DEN0056E)
  - Pin Control: v1.0.0 (SCMI v3.2 DEN0056E)
  - Power Domain: v3.0.1 (SCMI v3.2 DEN0056E)
  - NXP CPU: v1.0.0 (vendor extension)

Updated macros:
  - DT_SCMI_PROTOCOL_DATA_DEFINE
  - DT_SCMI_PROTOCOL_DEFINE
  - DT_SCMI_PROTOCOL_DEFINE_NODEV
  - DT_INST_SCMI_PROTOCOL_DEFINE

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

yongxu-wang15 committed:
drivers: firmware: scmi: add protocol version negotiate

Add SCMI protocol version negotiation to handle compatibility between
different agent and platform versions. The negotiation automatically
downgrades to supported version when needed and maintains backward
compatibility with legacy systems.

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

yongxu-wang15 committed:
drivers: firmware: scmi: force polling mode for message attributes query

During reboot operations, interrupts are disabled but k_is_pre_kernel()
returns false, causing SCMI communication to fail when using interrupt
mode. Force polling mode to ensure reliable operation in all contexts.

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

yongxu-wang15 commented on drivers/firmware/scmi/common.c:
```diff
@@ -0,0 +1,148 @@
+		return ret;
+	}
+
+	*attributes = reply_buffer.attributes;
```
done

yongxu-wang15 commented on include/zephyr/drivers/firmware/scmi/protocol.h:
```diff
@@ -126,4 +136,46 @@ int scmi_send_message(struct scmi_protocol *proto,
+ *
+ * @param proto Protocol instance
+ * @param version Pointer to store protocol version
+ * @retval 0 if successful
```
sure, its should be add space, done

yongxu-wang15 commented on include/zephyr/drivers/firmware/scmi/protocol.h:
```diff
@@ -126,4 +136,46 @@ int scmi_send_message(struct scmi_protocol *proto,
+ *
+ * @param proto Protocol instance
+ * @param desired_version Desired protocol version
+ * @retval 0 if successful
```
added it now

yongxu-wang15 commented on include/zephyr/drivers/firmware/scmi/protocol.h:
```diff
@@ -76,6 +76,17 @@ enum scmi_common_cmd {
+ * @struct scmi_protocol_data
+ * @brief Common protocol data structure with version info
+ */
+struct scmi_protocol_data {
```
thanks suggestion, its more better flow, I have updated it from your comments, and add all scmi protocol version

yongxu-wang15 commented:
Rebased to fix file confict, thanks

LaurentiuM1234 commented on drivers/firmware/scmi/CMakeLists.txt:
```diff
@@ -12,5 +12,6 @@ zephyr_library_sources_ifdef(CONFIG_ARM_SCMI_CLK_HELPERS clk.c)
 zephyr_library_sources_ifdef(CONFIG_ARM_SCMI_PINCTRL_HELPERS pinctrl.c)
 zephyr_library_sources_ifdef(CONFIG_ARM_SCMI_POWER_DOMAIN_HELPERS power.c)
 zephyr_library_sources_ifdef(CONFIG_ARM_SCMI_SYSTEM system.c)
+zephyr_library_sources_ifdef(CONFIG_ARM_SCMI_REBOOT reboot.c)
```
keep sorted alphabetically?

yongxu-wang15 replied:
sure, done

LaurentiuM1234 commented on drivers/firmware/scmi/system.c:
```diff
@@ -0,0 +1,76 @@
+
+	use_polling = k_is_pre_kernel();
+
+	ret = scmi_send_message(proto, &msg, &reply, use_polling);
```
use `k_is_pre_kernel()` directly?

LaurentiuM1234 reviewed (commented):
Looking good. Left some final nits below.

LaurentiuM1234 reviewed (approved):
LGTM, thank you for your contribution!

yongxu-wang15 commented:
ping: @dbaluta @dsseng, hi can you help review this pr again? I think the comments left before have been resolved, thanks

dsseng reviewed (commented):
No extra comments on my side, all good. However I do not know any other relevant reviewers to add, sorry
